### PR TITLE
WIP: Support fingerprint verification confirmation

### LINKED ISSUE #89829: Git: Support fingerprint verification confirmation
Issue Type: <b>Bug</b>

[Happens when git host hasn’t been accessed via ssh before.]

When I synchronize changes, VS Code reports:

```
Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

When I do a `git push` in a Terminal, this happens (which fixes the issue for VS Code):

```
% git push
The authenticity of host 'github.com (140.82.118.3)' can't be established.
RSA key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8.
Are you sure you want to continue connecting (yes/no)? yes
```

VS Code version: Code 1.41.1 (26076a4de974ead31f97692a0d32f90d735645c0, 2019-12-18T14:57:51.166Z)
OS version: Darwin x64 19.3.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-6267U CPU @ 2.90GHz (4 x 2900)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 2, 2|
|Memory (System)|16.00GB (0.97GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
spellright|ban|3.0.50
markdown-table-formatter|fcr|2.0.3
center-line|Jef|1.0.3
rewrap|stk|1.9.1

</details>

### PR DESCRIPTION
**Pull request by AlexStrNik**
This PR fixes #89829

This PR is a work in progress. I opened it to receive feedback from the maintainers. 

The main idea is to use `node-pty` to call git methods that can request fingerprint confirmation. So git will run in the interactive mode that allows confirm or decline verification programmatically

### CODE REVIEW

**Comment by AlexStrNik** [comment]
@joaomoreno can you guide me, please? Where is it best to place the function that asks the user confirmation via popup?

**Comment by joaomoreno** [comment]
@Tyriar Asking for a review, in case you see some issue in using `node-pty` this way.

**Comment by joaomoreno** [comment]
@AlexStrNik You have to provide some kind of user-input-provider as part of `PullOptions`.

**Review by Tyriar — COMMENTED** [review]
@joaomoreno seems ok, I'm a little concerned adding the dependency though as particularly on Windows node-pty has some issues. For example the `/.../i.test` checks that are done may not actually work as cursor movements may be included in the output data. Also sometimes node-pty just doesn't work, anti-virus blocking the binary is the biggest problem here. If you want to go this route I'd suggest using it as a fallback only if pull is blocked by fingerprint verification and not the regular way it happens.

**Comment by AlexStrNik** [comment]
It works). I know there are many things need to be refactored. But anyway it works. 
image

@Tyriar and @joaomoreno I once again asking you for a review.

File: extensions/git/src/main.ts [hunk 1]
```diff
@@ -193,11 +217,11 @@ async function checkGitVersion(info: IGit): Promise<void> {
 		return;
 	}
 
-	const update = localize('updateGit', "Update Git");
+	const update = localize('updateGit', 'Update Git');
 	const neverShowAgain = localize('neverShowAgain', "Don't Show Again");
 
 	const choice = await window.showWarningMessage(
-		localize('git20', "You seem to have git {0} installed. Code works best with git >= 2", info.version),
```
Comment (Tyriar) [inline extensions/git/src/main.ts:200]: These double quotes are intended when used within `localize`

**Comment by Tyriar** [comment]
The basic idea around node-pty's use looks good to me now as it's purely used as a fallback to fill in missing functionality.

**Comment by EmbeddedBacon** [comment]
Would like to see this also.  If I perform a clone of a repo VS Code fails out 

```
Looking for git in: C:\MyFiles\Apps\Git\bin\git.exe
Using git 2.26.2.windows.1 from C:\MyFiles\Apps\Git\bin\git.exe
> git clone ssh://git@bitbucket.xxxxx.com/pfw/yyyyy.git c:\MyFiles\Projects\Test\project --progress
Cloning into 'c:\MyFiles\Projects\Test\project'...
Host key verification failed.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

But if I try cloning from the terminal window I will see

```Are you sure you want to continue connecting (yes/no/[fingerprint])?```

**Comment by joaomoreno** [comment]
@AlexStrNik I've decided to close this for the following reasons:

- Parsing ssh output is brittle, and will break eventually
- This adds a lot of new code, which increases maintainability
- The feature is quite niche and had zero upvotes

Sorry about that and thanks for the hard work!

**Pull request closed without merging**